# Projeto 3: Segmentação de clientes no ecommerce

Realizar uma segmentação de clientes aplicando a metodologia RFM para um e-commerce.

O dataset utilizado neste projeto foi obtido desse projeto no [Kaggle](https://www.kaggle.com/datasets/datacertlaboratoria/projeto-3-segmentao-de-clientes-no-ecommerce).

In [1]:
import pandas as pd
import numpy as np


In [2]:
df = pd.read_csv("./dados/vendas-por-fatura.csv")

Realizando análise exploratória do dataset

In [3]:
df.head(50)

N° da fatura       Data da fatura  ID Cliente            País  Quantidade  \
0        548370   3/30/2021 16:14:00     15528.0  United Kingdom         123   
1        575767  11/11/2021 11:11:00     17348.0  United Kingdom         163   
2       C570727  10/12/2021 11:32:00     12471.0         Germany          -1   
3        549106    4/6/2021 12:08:00     17045.0  United Kingdom           1   
4        573112  10/27/2021 15:33:00     16416.0  United Kingdom         357   
5        576630   11/16/2021 8:38:00     13816.0         Germany          91   
6        538125   12/9/2020 15:46:00     18225.0  United Kingdom          16   
7        544354   2/18/2021 10:42:00     13489.0  United Kingdom          64   
8        546369   3/11/2021 11:41:00     15513.0  United Kingdom          10   
9        570651  10/11/2021 13:34:00     14911.0            EIRE          86   
10      C543628   2/10/2021 15:51:00     16558.0  United Kingdom          -5   
11       539736  12/21/2020 15:18:00         NaN  United Kingdom         143   
12       537823   12/8/2020 14:25:00         NaN  United Kingdom        1484   
13       575581  11/10/2021 11:53:00     12659.0          France         105   
14       580987   12/6/2021 16:35:00     17389.0  United Kingdom          96   
15       543754   2/11/2021 14:26:00     18159.0  United Kingdom         139   
16       542418   1/27/2021 17:39:00         NaN  United Kingdom         -17   
17      C553508   5/17/2021 13:44:00     14000.0  United Kingdom          -3   
18       563233   8/14/2021 15:56:00     17988.0  United Kingdom           1   
19       545417    3/2/2021 13:42:00     15622.0  United Kingdom         222   
20      C577693  11/21/2021 11:48:00     12628.0         Germany          -3   
21       550759   4/20/2021 12:09:00         NaN  United Kingdom          -7   
22      C547309   3/22/2021 11:17:00     17416.0  United Kingdom         -12   
23       556427   6/10/2021 13:23:00         NaN  United Kingdom          92   
24      C555404    6/2/2021 18:28:00     17231.0  United Kingdom          -2   
25       576571  11/15/2021 13:48:00     12950.0  United Kingdom         164   
26       550623   4/19/2021 13:56:00     13719.0  United Kingdom          78   
27       576083   11/14/2021 8:22:00     17404.0          Sweden        1060   
28       562417    8/4/2021 16:32:00         NaN  United Kingdom         526   
29       570733  10/12/2021 11:33:00     17190.0  United Kingdom         160   
30       537026   12/3/2020 16:35:00     12395.0         Belgium         528   
31       552837   5/11/2021 14:24:00     12901.0  United Kingdom          72   
32      C579925    12/1/2021 9:16:00     15544.0  United Kingdom          -3   
33       550535   4/19/2021 11:19:00     14211.0  United Kingdom          72   
34       555213    6/1/2021 12:22:00     13094.0  United Kingdom         144   
35       549923    4/13/2021 9:59:00     13138.0  United Kingdom          72   
36       575198    11/9/2021 9:47:00     12921.0  United Kingdom          19   
37      C549986   4/13/2021 16:03:00     18178.0  United Kingdom          -2   
38       569005   9/29/2021 18:30:00     16268.0  United Kingdom         465   
39       566307    9/12/2021 9:37:00     14307.0  United Kingdom         323   
40       539049  12/15/2020 16:21:00     15615.0  United Kingdom         364   
41       549684   4/11/2021 13:24:00         NaN  United Kingdom           1   
42       580068   12/1/2021 13:41:00     17321.0  United Kingdom           5   
43       542745   1/31/2021 16:47:00     15881.0  United Kingdom         154   
44       575525  11/10/2021 11:06:00     13983.0  United Kingdom          72   
45      C562884   8/10/2021 11:25:00     14056.0  United Kingdom          -3   
46      C560842   7/21/2021 12:25:00     14527.0  United Kingdom          -1   
47       542936    2/2/2021 11:10:00         NaN  United Kingdom         144   
48       565682    9/6/2021 10:42:00     14796.0  United Kingdom         467   
49 

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25953 entries, 0 to 25952
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   N° da fatura    25953 non-null  object 
 1   Data da fatura  25953 non-null  object 
 2   ID Cliente      22229 non-null  float64
 3   País            25953 non-null  object 
 4   Quantidade      25953 non-null  int64  
 5   Valor           25953 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 1.2+ MB


In [5]:
df.dtypes

N° da fatura       object
Data da fatura     object
ID Cliente        float64
País               object
Quantidade          int64
Valor              object
dtype: object

In [6]:
# Renomeando as colunas
df.columns = df.columns.str.replace(" ", "_")
df.columns

Index(['N°_da_fatura', 'Data_da_fatura', 'ID_Cliente', 'País', 'Quantidade',
       'Valor'],
      dtype='object')

In [7]:
df.shape

(25953, 6)

As faturas iniciadas com a letra 'C' 

Removendo as linhas das faturas que não possuem ID de clientes registrados, pois assim não é possível identificar qual o cliente e se ele não já foi anteriormente contabilizado.

In [8]:
df = df.dropna(subset=["ID_Cliente"])
df.shape

(22229, 6)

Foram removidos 3.724 registros sem ID (25953 linhas iniciais, resultando em 22229 linhas).

In [9]:
# Convertendo para tipo inteiro o ID
df['ID_Cliente'] = df['ID_Cliente'].astype(int)

In [10]:
df['Valor'] = df['Valor'].astype(str)
df.head()

N°_da_fatura       Data_da_fatura  ID_Cliente            País  Quantidade  \
0       548370   3/30/2021 16:14:00       15528  United Kingdom         123   
1       575767  11/11/2021 11:11:00       17348  United Kingdom         163   
2      C570727  10/12/2021 11:32:00       12471         Germany          -1   
3       549106    4/6/2021 12:08:00       17045  United Kingdom           1   
4       573112  10/27/2021 15:33:00       16416  United Kingdom         357   

    Valor  
0  229,33  
1  209,73  
2   -1,45  
3   39,95  
4  344,83

In [11]:
# Convertendo tipo da 'Data da Fatura'
df["Data_da_fatura"] = pd.to_datetime(df["Data_da_fatura"])
df.head()


N°_da_fatura      Data_da_fatura  ID_Cliente            País  Quantidade  \
0       548370 2021-03-30 16:14:00       15528  United Kingdom         123   
1       575767 2021-11-11 11:11:00       17348  United Kingdom         163   
2      C570727 2021-10-12 11:32:00       12471         Germany          -1   
3       549106 2021-04-06 12:08:00       17045  United Kingdom           1   
4       573112 2021-10-27 15:33:00       16416  United Kingdom         357   

    Valor  
0  229,33  
1  209,73  
2   -1,45  
3   39,95  
4  344,83

In [12]:
# Convertendo o tipo de 'Valor'
df[]


SyntaxError: invalid syntax (Temp/ipykernel_25304/2895650547.py, line 2)

In [ ]:
# Removendo os dados duplicados do dataset
df.drop_duplicates(inplace=True)

In [ ]:
# Classificando se a fatura é devolução ou não
df.loc[df['N°_da_fatura'].str.startswith('C'), 'Devolucao'] = "SIM"
df.loc[~df['N°_da_fatura'].str.startswith('C'), 'Devolucao'] = "NÃO"
df.head(20)

In [ ]:
# Removendo as faturas do tipo 'Devolução' do dataset

In [ ]:
# Ordenando ascendente pela Data da Fatura
df.sort_values(by=['Data_da_fatura'], inplace=True)
df.head(10)


N°_da_fatura      Data_da_fatura  ID_Cliente            País  \
9367        536365 2020-12-01 08:26:00       17850  United Kingdom   
18259       536366 2020-12-01 08:28:00       17850  United Kingdom   
11185       536368 2020-12-01 08:34:00       13047  United Kingdom   
6876        536367 2020-12-01 08:34:00       13047  United Kingdom   
8195        536369 2020-12-01 08:35:00       13047  United Kingdom   
505         536370 2020-12-01 08:45:00       12583          France   
12932       536371 2020-12-01 09:00:00       13748  United Kingdom   
2773        536372 2020-12-01 09:01:00       17850  United Kingdom   
20456       536373 2020-12-01 09:02:00       17850  United Kingdom   
24617       536374 2020-12-01 09:09:00       15100  United Kingdom   

       Quantidade   Valor  
9367           40  139,12  
18259          12   22,20  
11185          15   70,05  
6876           83  278,73  
8195            3   17,85  
505           449  855,86  
12932          80  204,00  
2773           12   22,20  
20456          88  259,86  
24617          32  350,40

In [ ]:
# Obtendo Ano e Mês
df['Ano_mes'] = df['Data_da_fatura'].dt.to_period('M')
df.head(10)


N°_da_fatura      Data_da_fatura  ID_Cliente            País  \
9367        536365 2020-12-01 08:26:00       17850  United Kingdom   
18259       536366 2020-12-01 08:28:00       17850  United Kingdom   
11185       536368 2020-12-01 08:34:00       13047  United Kingdom   
6876        536367 2020-12-01 08:34:00       13047  United Kingdom   
8195        536369 2020-12-01 08:35:00       13047  United Kingdom   
505         536370 2020-12-01 08:45:00       12583          France   
12932       536371 2020-12-01 09:00:00       13748  United Kingdom   
2773        536372 2020-12-01 09:01:00       17850  United Kingdom   
20456       536373 2020-12-01 09:02:00       17850  United Kingdom   
24617       536374 2020-12-01 09:09:00       15100  United Kingdom   

       Quantidade   Valor  Ano_mes  
9367           40  139,12  2020-12  
18259          12   22,20  2020-12  
11185          15   70,05  2020-12  
6876           83  278,73  2020-12  
8195            3   17,85  2020-12  
505           449  855,86  2020-12  
12932          80  204,00  2020-12  
2773           12   22,20  2020-12  
20456          88  259,86  2020-12  
24617          32  350,40  2020-12

In [ ]:
# Classificando se o cliente pertence ou não ao Reino Unido
df.loc[df['País']=="United Kingdom", 'Pertence_UK'] = "SIM"
df.loc[df['País']!="United Kingdom", 'Pertence_UK'] = "NÃO"
df.head(50)

N°_da_fatura      Data_da_fatura  ID_Cliente            País  \
9367        536365 2020-12-01 08:26:00       17850  United Kingdom   
18259       536366 2020-12-01 08:28:00       17850  United Kingdom   
11185       536368 2020-12-01 08:34:00       13047  United Kingdom   
6876        536367 2020-12-01 08:34:00       13047  United Kingdom   
8195        536369 2020-12-01 08:35:00       13047  United Kingdom   
505         536370 2020-12-01 08:45:00       12583          France   
12932       536371 2020-12-01 09:00:00       13748  United Kingdom   
2773        536372 2020-12-01 09:01:00       17850  United Kingdom   
20456       536373 2020-12-01 09:02:00       17850  United Kingdom   
24617       536374 2020-12-01 09:09:00       15100  United Kingdom   
25140       536375 2020-12-01 09:32:00       17850  United Kingdom   
1446        536376 2020-12-01 09:32:00       15291  United Kingdom   
6508        536377 2020-12-01 09:34:00       17850  United Kingdom   
24546       536378 2020-12-01 09:37:00       14688  United Kingdom   
660         536381 2020-12-01 09:41:00       15311  United Kingdom   
16673       536380 2020-12-01 09:41:00       17809  United Kingdom   
24816      C536379 2020-12-01 09:41:00       14527  United Kingdom   
8747        536382 2020-12-01 09:45:00       16098  United Kingdom   
21949      C536383 2020-12-01 09:49:00       15311  United Kingdom   
21244       536384 2020-12-01 09:53:00       18074  United Kingdom   
24139       536385 2020-12-01 09:56:00       17420  United Kingdom   
10263       536386 2020-12-01 09:57:00       16029  United Kingdom   
14830       536387 2020-12-01 09:58:00       16029  United Kingdom   
6927        536388 2020-12-01 09:59:00       16250  United Kingdom   
19929       536389 2020-12-01 10:03:00       12431       Australia   
5991        536390 2020-12-01 10:19:00       17511  United Kingdom   
14775      C536391 2020-12-01 10:24:00       17548  United Kingdom   
3973        536392 2020-12-01 10:29:00       13705  United Kingdom   
20755       536393 2020-12-01 10:37:00       13747  United Kingdom   
16119       536394 2020-12-01 10:39:00       13408  United Kingdom   
19864       536395 2020-12-01 10:47:00       13767  United Kingdom   
18246       536397 2020-12-01 10:51:00       17924  United Kingdom   
25338       536396 2020-12-01 10:51:00       17850  United Kingdom   
9596        536399 2020-12-01 10:52:00       17850  United Kingdom   
20807       536398 2020-12-01 10:52:00       13448  United Kingdom   
23237       536400 2020-12-01 10:53:00       13448  United Kingdom   
7968        536401 2020-12-01 11:21:00       15862  United Kingdom   
23089       536402 2020-12-01 11:22:00       15513  United Kingdom   
16746       536403 2020-12-01 11:27:00       12791     Netherlands   
19162       536404 2020-12-01 11:29:00       16218  United Kingdom   
4738        536405 2020-12-01 11:32:00       14045  United Kingdom   
21424       536406 2020-12-01 11:33:00       17850  United Kingdom   
20477       536407 2020-12-01 11:34:00       17850  United Kingdom   
16951       536408 2020-12-01 11:41:00       14307  United Kingdom   
6758        536409 2020-12-01 11:45:00       17908  United Kingdom   
25873       536412 2020-12-01 11:49:00       17920  United Kingdom   
2844        536415 2020-12-01 11:57:00       12838  United Kingdom   
24680       536416 2020-12-01 11:58:00       13255  United Kingdom   
16252       536420 2020-12-01 12:03:00       16583  United Kingdom   
13877       536423 2020-12-01 12:08:00       18085  United Kingdom   

       Quantidade    Valor  Ano_mes Pertence_UK  
9367           40   139,12  2020-12         SIM  
18259          12    22,20  2020-12         SIM  
11185          15    70,05  2020-12         SIM  
6876           83   278,73  2020-12         SIM  
8195            3    17,85  2020-12         SIM  
505           449   855,86  2020-12         NÃO  
12932          80   204,00  2020-12         SIM  
2773           12    22,20  2020-12

In [ ]:
df.head()

N°_da_fatura      Data_da_fatura  ID_Cliente            País  \
9367        536365 2020-12-01 08:26:00       17850  United Kingdom   
18259       536366 2020-12-01 08:28:00       17850  United Kingdom   
11185       536368 2020-12-01 08:34:00       13047  United Kingdom   
6876        536367 2020-12-01 08:34:00       13047  United Kingdom   
8195        536369 2020-12-01 08:35:00       13047  United Kingdom   

       Quantidade   Valor  Ano_mes Pertence_UK Devolucao  
9367           40  139,12  2020-12         SIM       NÃO  
18259          12   22,20  2020-12         SIM       NÃO  
11185          15   70,05  2020-12         SIM       NÃO  
6876           83  278,73  2020-12         SIM       NÃO  
8195            3   17,85  2020-12         SIM       NÃO

In [ ]:
df.dtypes

N°_da_fatura              object
Data_da_fatura    datetime64[ns]
ID_Cliente                 int32
País                      object
Quantidade                 int64
Valor                     object
Ano_mes                period[M]
Pertence_UK               object
Devolucao                 object
Total                     object
dtype: object

In [ ]:
df.drop(labels='Devolucao', axis=1, inplace=True)
df.head()

N°_da_fatura      Data_da_fatura  ID_Cliente            País  \
9367        536365 2020-12-01 08:26:00       17850  United Kingdom   
18259       536366 2020-12-01 08:28:00       17850  United Kingdom   
11185       536368 2020-12-01 08:34:00       13047  United Kingdom   
6876        536367 2020-12-01 08:34:00       13047  United Kingdom   
8195        536369 2020-12-01 08:35:00       13047  United Kingdom   

       Quantidade   Valor  Ano_mes Pertence_UK  \
9367           40  139,12  2020-12         SIM   
18259          12   22,20  2020-12         SIM   
11185          15   70,05  2020-12         SIM   
6876           83  278,73  2020-12         SIM   
8195            3   17,85  2020-12         SIM   

                                                   Total  
9367   139,12139,12139,12139,12139,12139,12139,12139,...  
18259  22,2022,2022,2022,2022,2022,2022,2022,2022,202...  
11185  70,0570,0570,0570,0570,0570,0570,0570,0570,057...  
6876   278,73278,73278,73278,73278,73278,73278,73278,...  
8195                                     17,8517,8517,85

In [ ]:
temp = df['Valor'][0]+df['Valor'][1]
temp

'229,33209,73'

Cliente  Data_ultima_compra  Recencia
275     12680 2021-12-09 12:50:00         0
587     13113 2021-12-09 12:49:00         0
2562    15804 2021-12-09 12:31:00         0
1067    13777 2021-12-09 12:25:00         0
3854    17581 2021-12-09 12:21:00         0
330     12748 2021-12-09 12:20:00         0
301     12713 2021-12-09 12:16:00         0
146     12526 2021-12-09 12:09:00         0
3215    16705 2021-12-09 12:08:00         0
2192    15311 2021-12-09 12:00:00         0

In [ ]:
print(df['ID_Cliente'].unique())

[17850 13047 12583 ... 13298 14569 12713]


In [ ]:
# Verificando se há valores nulos
temp = df.isnull().values.any()
temp


False

In [ ]:
# contando clientes únicos por ano/mês
df_coorte = pd.DataFrame(df.groupby(['Ano_mes'])['ID_Cliente'].nunique()).reset_index()
df_coorte

Ano_mes  ID_Cliente
0   2020-12         948
1   2021-01         783
2   2021-02         798
3   2021-03        1020
4   2021-04         899
5   2021-05        1079
6   2021-06        1051
7   2021-07         993
8   2021-08         980
9   2021-09        1302
10  2021-10        1425
11  2021-11        1711
12  2021-12         686

In [ ]:
# Criando coluna de Recencia 
df_recencia = df.groupby(by='ID_Cliente',
                        as_index=False)['Data_da_fatura'].max()
df_recencia.columns = ['Cliente', 'Data_ultima_compra']
ultima_compra = df_recencia['Data_ultima_compra'].max()
df_recencia['Recencia'] = df_recencia['Data_ultima_compra'].apply(
    lambda x: (ultima_compra - x).days)
df_recencia.head()

Cliente  Data_ultima_compra  Recencia
0    12346 2021-01-18 10:17:00       325
1    12347 2021-12-07 15:52:00         1
2    12348 2021-09-25 13:13:00        74
3    12349 2021-11-21 09:51:00        18
4    12350 2021-02-02 16:01:00       309

In [ ]:
df_frequencia = df.drop_duplicates().groupby(
    by=['ID_Cliente'], as_index=False)['Data_da_fatura'].count()
df_frequencia.columns = ['Cliente', 'Frequencia']
df_frequencia.head()

Cliente  Frequencia
0    12346           2
1    12347           7
2    12348           4
3    12349           1
4    12350           1

In [ ]:
df['Total'] = df['Valor']*df['Quantidade']
df_monetario = df.groupby(by='Cliente', as_index=False)['Total'].sum()
df_monetario.columns = ['Cliente', 'Monetario']
df_monetario.head()

KeyError: 'Cliente'